diff -u -N qmail-1.03-orig/Makefile qmail-1.03/Makefile
--- qmail-1.03-orig/Makefile	Mon Jun 15 04:53:16 1998
+++ qmail-1.03/Makefile	Thu Jun 25 19:28:36 1998
@@ -27,15 +27,6 @@
 	echo LD=\'`head -1 conf-ld`\' \
 	) > auto-ccld.sh
 
-auto-gid: \
-load auto-gid.o substdio.a error.a str.a fs.a
-	./load auto-gid substdio.a error.a str.a fs.a 
-
-auto-gid.o: \
-compile auto-gid.c subfd.h substdio.h substdio.h readwrite.h exit.h \
-scan.h fmt.h
-	./compile auto-gid.c
-
 auto-int: \
 load auto-int.o substdio.a error.a str.a fs.a
 	./load auto-int substdio.a error.a str.a fs.a 
@@ -60,15 +51,6 @@
 compile auto-str.c substdio.h readwrite.h exit.h
 	./compile auto-str.c
 
-auto-uid: \
-load auto-uid.o substdio.a error.a str.a fs.a
-	./load auto-uid substdio.a error.a str.a fs.a 
-
-auto-uid.o: \
-compile auto-uid.c subfd.h substdio.h substdio.h readwrite.h exit.h \
-scan.h fmt.h
-	./compile auto-uid.c
-
 auto_break.c: \
 auto-str conf-break
 	./auto-str auto_break \
@@ -110,22 +92,8 @@
 compile auto_split.c
 	./compile auto_split.c
 
-auto_uids.c: \
-auto-uid auto-gid conf-users conf-groups
-	( ./auto-uid auto_uida `head -1 conf-users` \
-	&&./auto-uid auto_uidd `head -2 conf-users | tail -1` \
-	&&./auto-uid auto_uidl `head -3 conf-users | tail -1` \
-	&&./auto-uid auto_uido `head -4 conf-users | tail -1` \
-	&&./auto-uid auto_uidp `head -5 conf-users | tail -1` \
-	&&./auto-uid auto_uidq `head -6 conf-users | tail -1` \
-	&&./auto-uid auto_uidr `head -7 conf-users | tail -1` \
-	&&./auto-uid auto_uids `head -8 conf-users | tail -1` \
-	&&./auto-gid auto_gidq `head -1 conf-groups` \
-	&&./auto-gid auto_gidn `head -2 conf-groups | tail -1` \
-	) > auto_uids.c.tmp && mv auto_uids.c.tmp auto_uids.c
-
 auto_uids.o: \
-compile auto_uids.c
+compile auto_uids.c auto_uids.h
 	./compile auto_uids.c
 
 auto_usera.c: \
@@ -740,17 +708,16 @@
 	./compile idedit.c
 
 install: \
-load install.o fifo.o hier.o auto_qmail.o auto_split.o auto_uids.o \
+load install.o fifo.o hier.o auto_split.o auto_uids.o \
 strerr.a substdio.a open.a error.a str.a fs.a
-	./load install fifo.o hier.o auto_qmail.o auto_split.o \
+	./load install fifo.o hier.o auto_split.o \
 	auto_uids.o strerr.a substdio.a open.a error.a str.a fs.a 
 
 install-big: \
-load install-big.o fifo.o install.o auto_qmail.o auto_split.o \
-auto_uids.o strerr.a substdio.a open.a error.a str.a fs.a
-	./load install-big fifo.o install.o auto_qmail.o \
-	auto_split.o auto_uids.o strerr.a substdio.a open.a error.a \
-	str.a fs.a 
+load install-big.o fifo.o install.o auto_split.o auto_uids.o \
+strerr.a substdio.a open.a error.a str.a fs.a
+	./load install-big fifo.o install.o auto_split.o auto_uids.o \
+	strerr.a substdio.a open.a error.a str.a fs.a 
 
 install-big.o: \
 compile install-big.c auto_qmail.h auto_split.h auto_uids.h fmt.h \
@@ -763,9 +730,9 @@
 	./compile install.c
 
 instcheck: \
-load instcheck.o fifo.o hier.o auto_qmail.o auto_split.o auto_uids.o \
+load instcheck.o fifo.o hier.o auto_split.o auto_uids.o \
 strerr.a substdio.a error.a str.a fs.a
-	./load instcheck fifo.o hier.o auto_qmail.o auto_split.o \
+	./load instcheck fifo.o hier.o auto_split.o \
 	auto_uids.o strerr.a substdio.a error.a str.a fs.a 
 
 instcheck.o: \
@@ -808,7 +775,7 @@
 forward preline condredirect bouncesaying except maildirmake \
 maildir2mbox maildirwatch qail elq pinq idedit install-big install \
 instcheck home home+df proc proc+df binm1 binm1+df binm2 binm2+df \
-binm3 binm3+df
+binm3 binm3+df make-owners
 
 load: \
 make-load warn-auto.sh systype
@@ -918,6 +885,15 @@
 	cat auto-ccld.sh make-makelib.sh > make-makelib
 	chmod 755 make-makelib
 
+make-owners: \
+make-owners.head conf-users conf-groups
+	cat make-owners.head >make-owners
+	for num in a d l o p q r s; do read name; \
+		echo checkuid $$num $$name; done <conf-users >>make-owners
+	for num in q n; do read name; \
+		echo checkgid $$num $$name; done <conf-groups >>make-owners
+	chmod +x make-owners
+
 makelib: \
 make-makelib warn-auto.sh systype
 	( cat warn-auto.sh; ./make-makelib "`cat systype`" ) > \
@@ -1047,7 +1023,7 @@
 	chmod 755 proc+df
 
 prot.o: \
-compile prot.c hasshsgr.h prot.h
+compile prot.c hasshsgr.h prot.h auto_uids.h
 	./compile prot.c
 
 qail: \
@@ -1201,10 +1177,9 @@
 load qmail-lspawn.o spawn.o prot.o slurpclose.o coe.o sig.a wait.a \
 case.a cdb.a fd.a open.a stralloc.a alloc.a substdio.a error.a str.a \
 fs.a auto_qmail.o auto_uids.o auto_spawn.o
-	./load qmail-lspawn spawn.o prot.o slurpclose.o coe.o \
+	./load qmail-lspawn spawn.o prot.o slurpclose.o coe.o auto_uids.o \
 	sig.a wait.a case.a cdb.a fd.a open.a stralloc.a alloc.a \
-	substdio.a error.a str.a fs.a auto_qmail.o auto_uids.o \
-	auto_spawn.o 
+	substdio.a error.a str.a fs.a auto_qmail.o auto_spawn.o 
 
 qmail-lspawn.0: \
 qmail-lspawn.8
@@ -1422,10 +1397,10 @@
 load qmail-queue.o triggerpull.o fmtqfn.o now.o date822fmt.o \
 datetime.a seek.a ndelay.a open.a sig.a alloc.a substdio.a error.a \
 str.a fs.a auto_qmail.o auto_split.o auto_uids.o
-	./load qmail-queue triggerpull.o fmtqfn.o now.o \
+	./load qmail-queue triggerpull.o fmtqfn.o now.o auto_uids.o \
 	date822fmt.o datetime.a seek.a ndelay.a open.a sig.a \
 	alloc.a substdio.a error.a str.a fs.a auto_qmail.o \
-	auto_split.o auto_uids.o 
+	auto_split.o
 
 qmail-queue.0: \
 qmail-queue.8
@@ -1464,10 +1439,9 @@
 load qmail-rspawn.o spawn.o tcpto_clean.o now.o coe.o sig.a open.a \
 seek.a lock.a wait.a fd.a stralloc.a alloc.a substdio.a error.a str.a \
 auto_qmail.o auto_uids.o auto_spawn.o
-	./load qmail-rspawn spawn.o tcpto_clean.o now.o coe.o \
+	./load qmail-rspawn spawn.o tcpto_clean.o now.o coe.o auto_uids.o \
 	sig.a open.a seek.a lock.a wait.a fd.a stralloc.a alloc.a \
-	substdio.a error.a str.a auto_qmail.o auto_uids.o \
-	auto_spawn.o 
+	substdio.a error.a str.a auto_qmail.o auto_spawn.o 
 
 qmail-rspawn.0: \
 qmail-rspawn.8
@@ -1557,8 +1531,10 @@
 	./compile qmail-smtpd.c
 
 qmail-start: \
-load qmail-start.o prot.o fd.a auto_uids.o
-	./load qmail-start prot.o fd.a auto_uids.o 
+load qmail-start.o prot.o fd.a auto_uids.o auto_qmail.o \
+	open_read.o substdio.a error.a
+	./load qmail-start prot.o auto_uids.o fd.a \
+	auto_qmail.o open_read.o substdio.a str.a error.a
 
 qmail-start.0: \
 qmail-start.8
@@ -1792,7 +1768,7 @@
 proc+df.sh binm1.sh binm2.sh binm3.sh binm1+df.sh binm2+df.sh \
 binm3+df.sh find-systype.sh make-compile.sh make-load.sh \
 make-makelib.sh trycpp.c warn-auto.sh auto-str.c auto-int.c \
-auto-int8.c auto-gid.c auto-uid.c hier.c install.c instcheck.c \
+auto-int8.c hier.c install.c instcheck.c \
 install-big.c alloc.3 alloc.h alloc.c alloc_re.c case.3 case.h \
 case_diffb.c case_diffs.c case_lowerb.c case_lowers.c case_starts.c \
 cdb.3 cdb.h cdb_hash.c cdb_seek.c cdb_unpack.c cdbmake.h \
diff -u -N qmail-1.03-orig/TARGETS qmail-1.03/TARGETS
--- qmail-1.03-orig/TARGETS	Mon Jun 15 04:53:16 1998
+++ qmail-1.03/TARGETS	Thu Jun 25 18:10:53 1998
@@ -150,11 +150,6 @@
 cdb_unpack.o
 cdb_seek.o
 cdb.a
-auto-uid.o
-auto-uid
-auto-gid.o
-auto-gid
-auto_uids.c
 auto_uids.o
 qmail-lspawn
 qmail-getpw.o
@@ -385,3 +380,4 @@
 man
 setup
 check
+make-owners
diff -u -N qmail-1.03-orig/auto_uids.c qmail-1.03/auto_uids.c
--- qmail-1.03-orig/auto_uids.c	Wed Dec 31 18:00:00 1969
+++ qmail-1.03/auto_uids.c	Thu Jun 25 18:09:46 1998
@@ -0,0 +1,108 @@
+#include <stdlib.h>
+#include <sys/stat.h>
+#include "exit.h"
+#include "subfd.h"
+#include "substdio.h"
+#include "auto_qmail.h"
+#include "auto_uids.h"
+
+/* These are offsets from the end of either uid_files or gid_files */
+
+int auto_uida = -8;
+int auto_uidd = -7;
+int auto_uidl = -6;
+int auto_uido = -5;
+int auto_uidp = -4;
+int auto_uidq = -3;
+int auto_uidr = -2;
+int auto_uids = -1;
+
+int auto_gidn = -2;
+int auto_gidq = -1;
+
+#define uid_table_size 8
+#define gid_table_size 2
+
+struct file_ref { const char *name; int *var; };
+
+static struct file_ref uid_files[uid_table_size] = {
+	{ "uida", &auto_uida },
+	{ "uidd", &auto_uidd },
+	{ "uidl", &auto_uidl },
+	{ "uido", &auto_uido },
+	{ "uidp", &auto_uidp },
+	{ "uidq", &auto_uidq },
+	{ "uidr", &auto_uidr },
+	{ "uids", &auto_uids }
+};
+
+static struct file_ref gid_files[gid_table_size] = {
+	{ "gidn", &auto_gidn },
+	{ "gidq", &auto_gidq }
+};
+
+static int stat_control_file(name, buf) char* name; struct stat* buf;
+{
+  int result;
+  char* file = malloc(strlen(auto_qmail) + strlen(name) + 10);
+  if(file == 0)
+   {
+    substdio_putsflush(subfderr,"fatal: unable to allocate memory\n");
+    _exit(111);
+   }
+  strcpy(file, auto_qmail);
+  strcat(file, "/owners/");
+  strcat(file, name);
+  result = stat(file, buf);
+  free(file);
+  return result;
+}
+
+static int stat_uid_file(ref) struct file_ref* ref;
+{
+  struct stat statbuf;
+  if(stat_control_file(ref->name, &statbuf) == -1)
+   {
+    substdio_puts(subfderr,"fatal: unable to stat uid control file '");
+    substdio_puts(subfderr,ref->name);
+    substdio_puts(subfderr,"'\n");
+    substdio_flush(subfderr);
+    _exit(111);
+   }
+  return *(ref->var) = statbuf.st_uid;
+}
+
+static int stat_gid_file(ref) struct file_ref* ref;
+{
+  struct stat statbuf;
+  if(stat_control_file(ref->name, &statbuf) == -1)
+   {
+    substdio_puts(subfderr,"fatal: unable to stat gid control file '");
+    substdio_puts(subfderr,ref->name);
+    substdio_puts(subfderr,"'\n");
+    substdio_flush(subfderr);
+    _exit(111);
+   }
+  return *(ref->var) = statbuf.st_gid;
+}
+
+int get_uid(id) int id;
+{
+  if(id >= 0)
+    return id;
+  else if(id < -uid_table_size)
+    return -1;
+  else
+    return stat_uid_file(&uid_files[uid_table_size+id]);
+}
+
+int get_gid(id) int id;
+{
+  if(id >= 0)
+    return id;
+  else if(id < -gid_table_size)
+    return -1;
+  else
+    return stat_gid_file(&gid_files[gid_table_size+id]);
+}
+
diff -u -N qmail-1.03-orig/auto_uids.h qmail-1.03/auto_uids.h
--- qmail-1.03-orig/auto_uids.h	Mon Jun 15 04:53:16 1998
+++ qmail-1.03/auto_uids.h	Thu Jun 25 18:09:46 1998
@@ -13,4 +13,7 @@
 extern int auto_gidn;
 extern int auto_gidq;
 
+extern int get_uid();
+extern int get_gid();
+
 #endif
diff -u -N qmail-1.03-orig/hier.c qmail-1.03/hier.c
--- qmail-1.03-orig/hier.c	Mon Jun 15 04:53:16 1998
+++ qmail-1.03/hier.c	Thu Jun 25 19:28:36 1998
@@ -4,6 +4,15 @@
 #include "fmt.h"
 #include "fifo.h"
 
+/* Fool auto_uids.o into pulling the uid/gid files from the subdirectory
+ * "owners" in the current directory.  This requires that the command
+ * "./make-owners ." be executed in the source directory before running
+ * qmail-hier.  This unfortunately hoses up the standard install process
+ * (it will install into the current directory instead of /var/qmail),
+ * and so it requires the install-path patch for proper operation.
+ */
+char auto_qmail[] = ".";
+
 char buf[100 + FMT_ULONG];
 
 void dsplit(base,uid,mode)
@@ -29,10 +38,23 @@
 
 void hier()
 {
+  get_uid(auto_uida);
+  get_uid(auto_uidd);
+  get_uid(auto_uidl);
+  get_uid(auto_uido);
+  get_uid(auto_uidp);
+  get_uid(auto_uidq);
+  get_uid(auto_uidr);
+  get_uid(auto_uids);
+
+  get_gid(auto_gidq);
+  get_gid(auto_gidn);
+
   h(auto_qmail,auto_uido,auto_gidq,0755);
 
   d(auto_qmail,"control",auto_uido,auto_gidq,0755);
   d(auto_qmail,"users",auto_uido,auto_gidq,0755);
+  d(auto_qmail,"owners",auto_uido,auto_gidq,0755);
   d(auto_qmail,"bin",auto_uido,auto_gidq,0755);
   d(auto_qmail,"boot",auto_uido,auto_gidq,0755);
   d(auto_qmail,"doc",auto_uido,auto_gidq,0755);
@@ -143,6 +165,7 @@
   c(auto_qmail,"bin","qail",auto_uido,auto_gidq,0755);
   c(auto_qmail,"bin","elq",auto_uido,auto_gidq,0755);
   c(auto_qmail,"bin","pinq",auto_uido,auto_gidq,0755);
+  c(auto_qmail,"bin","make-owners",auto_uido,auto_gidq,0755);
 
   c(auto_qmail,"man/man5","addresses.5",auto_uido,auto_gidq,0644);
   c(auto_qmail,"man/cat5","addresses.0",auto_uido,auto_gidq,0644);
diff -u -N qmail-1.03-orig/install-big.c qmail-1.03/install-big.c
--- qmail-1.03-orig/install-big.c	Mon Jun 15 04:53:16 1998
+++ qmail-1.03/install-big.c	Thu Jun 25 19:28:36 1998
@@ -4,6 +4,15 @@
 #include "fmt.h"
 #include "fifo.h"
 
+/* Fool auto_uids.o into pulling the uid/gid files from the subdirectory
+ * "owners" in the current directory.  This requires that the command
+ * "./make-owners ." be executed in the source directory before running
+ * qmail-hier.  This unfortunately hoses up the standard install, and so
+ * (it will install into the current directory instead of /var/qmail),
+ * and so it requires the install-path patch for proper operation.
+ */
+char auto_qmail[] = ".";
+
 char buf[100 + FMT_ULONG];
 
 void dsplit(base,uid,mode)
@@ -29,6 +38,18 @@
 
 void hier()
 {
+  get_uid(auto_uida);
+  get_uid(auto_uidd);
+  get_uid(auto_uidl);
+  get_uid(auto_uido);
+  get_uid(auto_uidp);
+  get_uid(auto_uidq);
+  get_uid(auto_uidr);
+  get_uid(auto_uids);
+
+  get_gid(auto_gidq);
+  get_gid(auto_gidn);
+
   h(auto_qmail,auto_uido,auto_gidq,0755);
 
   d(auto_qmail,"control",auto_uido,auto_gidq,0755);
diff -u -N qmail-1.03-orig/make-owners.head qmail-1.03/make-owners.head
--- qmail-1.03-orig/make-owners.head	Wed Dec 31 18:00:00 1969
+++ qmail-1.03/make-owners.head	Thu Jun 25 18:09:46 1998
@@ -0,0 +1,34 @@
+#!/bin/sh
+
+set -e
+dir=$1
+if ! [ -d $dir ]
+then
+	echo "usage: make-owners directory"
+	exit 1
+fi
+dir=$dir/owners
+if ! [ -d $dir ]
+then
+	echo "make-owners: directory '$dir' does not exist, creating."
+	mkdir $dir
+fi
+
+checkuid() {
+  file=$dir/uid$1
+  if ! [ -e $file ]; then
+    echo "make-owners: UID file '$file' does not exist, creating."
+    touch $file
+  fi
+  chown $2 $file
+}
+
+checkgid() {
+  file=$dir/gid$1
+  if ! [ -e $file ]; then
+    echo "make-owners: GID file '$file' does not exist, creating."
+    touch $file
+  fi
+  chgrp $2 $file
+}
+
diff -u -N qmail-1.03-orig/prot.c qmail-1.03/prot.c
--- qmail-1.03-orig/prot.c	Mon Jun 15 04:53:16 1998
+++ qmail-1.03/prot.c	Thu Jun 25 18:09:46 1998
@@ -1,10 +1,12 @@
 #include "hasshsgr.h"
 #include "prot.h"
+#include "auto_uids.h"
 
 /* XXX: there are more portability problems here waiting to leap out at me */
 
 int prot_gid(gid) int gid;
 {
+  gid = get_gid(gid);
 #ifdef HASSHORTSETGROUPS
   short x[2];
   x[0] = gid; x[1] = 73; /* catch errors */
@@ -17,5 +19,6 @@
 
 int prot_uid(uid) int uid;
 {
+  uid = get_uid(uid);
   return setuid(uid);
 }
diff -u -N qmail-1.03-orig/qmail-queue.c qmail-1.03/qmail-queue.c
--- qmail-1.03-orig/qmail-queue.c	Mon Jun 15 04:53:16 1998
+++ qmail-1.03/qmail-queue.c	Thu Jun 25 18:10:00 1998
@@ -73,11 +73,11 @@
  i = fmt_str(s,"Received: (qmail "); len += i; if (s) s += i;
  i = fmt_ulong(s,mypid); len += i; if (s) s += i;
  i = fmt_str(s," invoked "); len += i; if (s) s += i;
- if (uid == auto_uida)
+ if (uid == get_uid(auto_uida))
   { i = fmt_str(s,"by alias"); len += i; if (s) s += i; }
- else if (uid == auto_uidd)
+ else if (uid == get_uid(auto_uidd))
   { i = fmt_str(s,"from network"); len += i; if (s) s += i; }
- else if (uid == auto_uids)
+ else if (uid == get_uid(auto_uids))
   { i = fmt_str(s,"for bounce"); len += i; if (s) s += i; }
  else
   {
diff -u -N qmail-1.03-orig/qmail-showctl.c qmail-1.03/qmail-showctl.c
--- qmail-1.03-orig/qmail-showctl.c	Mon Jun 15 04:53:16 1998
+++ qmail-1.03/qmail-showctl.c	Thu Jun 25 22:24:17 1998
@@ -164,27 +164,27 @@
   substdio_puts(subfdout,".\n");
 
   substdio_puts(subfdout,"user ids: ");
-  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) auto_uida));
+  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) get_uid(auto_uida)));
   substdio_puts(subfdout,", ");
-  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) auto_uidd));
+  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) get_uid(auto_uidd)));
   substdio_puts(subfdout,", ");
-  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) auto_uidl));
+  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) get_uid(auto_uidl)));
   substdio_puts(subfdout,", ");
-  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) auto_uido));
+  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) get_uid(auto_uido)));
   substdio_puts(subfdout,", ");
-  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) auto_uidp));
+  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) get_uid(auto_uidp)));
   substdio_puts(subfdout,", ");
-  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) auto_uidq));
+  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) get_uid(auto_uidq)));
   substdio_puts(subfdout,", ");
-  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) auto_uidr));
+  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) get_uid(auto_uidr)));
   substdio_puts(subfdout,", ");
-  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) auto_uids));
+  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) get_uid(auto_uids)));
   substdio_puts(subfdout,".\n");
 
   substdio_puts(subfdout,"group ids: ");
-  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) auto_gidn));
+  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) get_gid(auto_gidn)));
   substdio_puts(subfdout,", ");
-  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) auto_gidq));
+  substdio_put(subfdout,num,fmt_ulong(num,(unsigned long) get_gid(auto_gidq)));
   substdio_puts(subfdout,".\n");
 
   if (chdir(auto_qmail) == -1) {
diff -u -N qmail-1.03-orig/spawn.c qmail-1.03/spawn.c
--- qmail-1.03-orig/spawn.c	Mon Jun 15 04:53:16 1998
+++ qmail-1.03/spawn.c	Thu Jun 25 18:10:00 1998
@@ -110,7 +110,7 @@
   { close(fdmess); err("Zqmail-spawn unable to fstat message. (#4.3.0)\n"); return; }
  if ((st.st_mode & S_IFMT) != S_IFREG)
   { close(fdmess); err("ZSorry, message has wrong type. (#4.3.5)\n"); return; }
- if (st.st_uid != auto_uidq) /* aaack! qmailq has to be trusted! */
+ if (st.st_uid != get_uid(auto_uidq)) /* aaack! qmailq has to be trusted! */
   /* your security is already toast at this point. damage control... */
   { close(fdmess); err("ZSorry, message has wrong owner. (#4.3.5)\n"); return; }
 
